packages/phone-number-privacy/common/src/utils/sequential-delay.ts

```diff
@@ -0,0 +1,80 @@
+import { SequentialDelayDomain, SequentialDelayStage } from '../interfaces'
+interface IndexedSequentialDelayStage extends SequentialDelayStage {
+  // The stage's index in the stage array
+  index: number
+  // The attempt number at which the stage begins
+  start: number
+}
+export interface SequentialDelayResult {
+  accepted: boolean
+  state?: SequentialDelayState
+}
+interface SequentialDelayState {
+  // Timestamp used for deciding when the next request will be accepted.
+  timer: number
+  // Number of queries that have been accepted for the SequentialDelayDomain instance.
+  counter: number
+}
+
+export const checkSequentialDelay = (
+  domain: SequentialDelayDomain,
+  attemptTime: number,
+  state?: SequentialDelayState
+): SequentialDelayResult => {
+  const counter = state?.counter ?? 0
+  const timer = state?.timer ?? 0
```

Comment:
(Just suggesting comments as I read through that would help me understand this code)
```suggestion
  // If no state is available (i.e. this is the first request against the domain) use the initial state.
  const counter = state?.counter ?? 0
  const timer = state?.timer ?? 0
```